MemoryCache: Size calculation with expired Entries

Bernd-Broesel opened the pull request:
Trying to add an already expired entry to the cache no longer
permanently increases the cache size

Fix #36037

Commit:
MemoryCache: Size calculation with expired Entries

Trying to add an already expired entry to the cache no longer
permanently increases the cache size

Fix #36037

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=39189) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=39189) before we can accept your contribution.<br/><br/>:x: hiersekornc [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=39189)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=39189) it.</sub>

Commit:
MemoryCache: Size calculation with expired Entries

Follow up: Only decrease cache size when a cache size limit actually exists

ericstj commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs:
```diff
@@ -341,6 +341,23 @@ public async Task ExpiringEntryDecreasesCacheSize()
         }
 
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/36037")]
```
Please remove `ActiveIssue`.  This attribute is used to disable a test against an issue when the test is failing.   You don't need that here.
```suggestion
```

hiersekornc replied:
Hi Eric, thanks for your review. The attribute has been removed. Additionally I have added a test for an expired token.

ericstj reviewed (commented):
Looks good, just one small suggestion.

Commit:
MemoryCache: Size calculation with expired Entries

Added a test for entries with expired tokens.
Removed ActiveIssue attribute from previous test.

ericstj commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs:
```diff
@@ -341,6 +341,39 @@ public async Task ExpiringEntryDecreasesCacheSize()
+        [Fact]
+        public void TryingToAddEntryWithExpiredTokenDoesNotIncreaseCacheSize()
+        {
+            var cache = new MemoryCache(new MemoryCacheOptions {SizeLimit = 10});
```
nit: formatting
```suggestion
            var cache = new MemoryCache(new MemoryCacheOptions { SizeLimit = 10 });
```

ericstj commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs:
```diff
@@ -341,6 +341,39 @@ public async Task ExpiringEntryDecreasesCacheSize()
+        public void TryingToAddExpiredEntryDoesNotIncreaseCacheSize()
+        {
+            var testClock = new TestClock();
+            var cache = new MemoryCache(new MemoryCacheOptions {Clock = testClock, SizeLimit = 10});
```
nit: formatting
```suggestion
            var cache = new MemoryCache(new MemoryCacheOptions { Clock = testClock, SizeLimit = 10 });
```

ericstj reviewed (approved):
LGTM with minor code formatting suggestions.  Please change from draft->official PR.   I'd also like a review from extensions-Caching owners @maryamariyan @eerhardt

maryamariyan commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs:
```diff
@@ -341,6 +341,39 @@ public async Task ExpiringEntryDecreasesCacheSize()
+            var testExpirationToken = new TestExpirationToken { HasChanged = true };
+            var entryOptions = new MemoryCacheEntryOptions
+            {
+                Size = 5, ExpirationTokens = { testExpirationToken }
```
nit: formatting
```suggestion
                Size = 5,
                ExpirationTokens = { testExpirationToken }
```

Commit:
MemoryCache: Size calculation with expired Entries

Fixed formatting

hiersekornc commented:
Formatting suggestions have been applied.

maryamariyan commented:
@hiersekornc thank you for your contribution!
